rafthttp/stream.go

```
@@ -0,0 +1,346 @@
+		case <-cw.stopc:
+			reset()
+			close(cw.done)
+			return
+		}
+	}
+}
+
+func (cw *streamWriter) isWorking() bool {
+	cw.mu.Lock()
+	defer cw.mu.Unlock()
+	return cw.working
+}
+
+func (cw *streamWriter) attach(conn *outgoingConn) bool {
+	select {
+	case cw.connc <- conn:
+		return true
+	case <-cw.done:
+		return false
+	}
+}
+
+func (cw *streamWriter) stop() {
+	close(cw.stopc)
+	<-cw.done
+}
+
+// streamReader is a long-running go-routine that dials to the remote stream
+// endponit and reads messages from the response body returned.
+type streamReader struct {
+	tr       http.RoundTripper
+	u        string
+	t        streamType
+	from, to types.ID
+	cid      types.ID
+	recvc    chan<- raftpb.Message
+
+	mu         sync.Mutex
+	msgAppTerm uint64
+	req        *http.Request
+	closer     io.Closer
+	stopc      chan struct{}
+	done       chan struct{}
+}
+
+func startStreamReader(tr http.RoundTripper, u string, t streamType, from, to, cid types.ID, recvc chan<- raftpb.Message) *streamReader {
+	r := &streamReader{
+		tr:    tr,
+		u:     u,
+		t:     t,
+		from:  from,
+		to:    to,
+		cid:   cid,
+		recvc: recvc,
+		stopc: make(chan struct{}),
+		done:  make(chan struct{}),
+	}
+	go r.run()
+	return r
+}
+
+func (cr *streamReader) run() {
+	for {
+		err := cr.dialAndDecode()
+		if err != nil {
+			log.Printf("rafthttp: %v", err)
+		}
+		select {
+		case <-time.After(100 * time.Millisecond):
+		case <-cr.stopc:
+			close(cr.done)
+			return
+		}
+	}
+}
+
+func (cr *streamReader) dialAndDecode() error {
+	var dec decoder
+	rc, err := cr.dial()
+	if err != nil {
+		return fmt.Errorf("rafthttp: roundtripping error: %v", err)
+	}
+	cr.mu.Lock()
+	switch cr.t {
+	case streamTypeMsgApp:
+		dec = &msgAppDecoder{r: rc, local: cr.from, remote: cr.to, term: cr.msgAppTerm}
+	case streamTypeMessage:
+		dec = &messageDecoder{r: rc}
+	default:
+		log.Panicf("rafthttp: unhandled stream type %s", cr.t)
+	}
+	cr.closer = rc
+	cr.mu.Unlock()
+
+	for {
+		m, err := dec.decode()
```

why not return the error and let the outside to parse the error?